policies/codingstyle-updates.txt

```diff
@@ -0,0 +1,70 @@
+A summary of the discussion thread so far.  Not surprisingly, it's all about
+the whitespace. :)
+
+The descriptions here were written to be understandable stand-alone.  Once we
+come to a conclusion, we'll wordsmith them into the coding style.
+
+Do not put a space after sizeof; do use parens:
+
+	sizeof(foo)
+
+When breaking long lines, if there are Boolean conditionals, put them at the
+start of the line.  Consider doing this consistently, and don't merge even if
+they fit.  Here is a bad example, it should be three lines:
+
+	some_long_condition()
+	&& short1() && short2()
+
+Related conditions should be on one line if they fit, else given an extra
+level of indent
+
+	some_long_condition()
+	    && (short1() || short2())
+	    && some_other_flag != 0
+
+If the expression for an if statement does not fit, indent the continuation
+lines an extra level (this is not necessary for an else-if)
+
+	if (this_is_true()
+		&& !that_is_false()) {
+	    code();
+	    ...
+
+Try not to break long lines within a function call, but if you have to, indent
+the rest of the parameter list to be after the function's opening paren.
+If the open paren is "far to the right," however, indent less and try to
+avoid a column of text.
+
+Remember a blank line after variable declarations (even local ones).
+
+Treat a single-statement with comment as if it were multi-line and use curly
+braces. Note that this could end up having "cascading curly" effects.
+
+	if (test()) {
+	    /* already alerted */
+	    close();
+	}
+
+Arguments inside macro expansions should be parenthesized, except when
+using the stringify or concatenation operators (# and ## respectively).
+The following examples show examples of this, and proper spacing.
+
+	#define foo(a, b) ((a) * (b))
+	#define quote(a) #a
+	#define join(a, b) a ## quote(a)
+
+Free routines should properly handle null; don't check for null before calling
+a free routine.
+
+Argument names in function definition should match the declaration, but you
+can use "unused_" as a prefix in the definition for unused arguments.
+
+When possible, use size_t for sizes of things (including array indicies)
+
+Prefer ossl_assert over assert. Do not use OPENSSL_assert() in libcrypto or
+libssl.

```

Comment:
As you might know I'm in favour of using as little lingo as possible. With this in mind I'd argue that ossl_assert and assert are equally appropriate, just serve different purposes. assert has rather documentation purpose, i.e. is to be used to document assumptions/boundary conditions, not unlikely static. ossl_assert on the other hand is meant to be used rather with `if(!ossl_assert()) handler()`, i.e. when there is something that can be done at run-time [in non-debug build] when assertion doesn't hold. Or in other words, if there is no actual need for ossl_assert I wouldn't be opposed to use of assert. Because it's standard and recognized construct. In other words I'd argue against unconditional ban of assert.

Reply:
The difference is that ossl_assert returns a value in non-debug environments while assert is void.  That makes
```
if (ossl_assert(...
```
possible.  It's used nearly 100 times in openssl code.

Reply:
> The difference is that ossl_assert returns a value in non-debug environments while assert is void.

This doesn't contradict what I'm saying, rather contrary. They are different and as such have different applications. And in such case you can't really say "prefer one over another"...

Reply:
I think there could be a tendency to use assert where ossl_assert might be more appropriate. So rather than saying prefer one over the other, perhaps we should explain when you should use each one?

Reply:
I thought we didn't want crashes, and that if you had an assert, there *had* to be a runtime test and error return, and that this led to ossl_assert which combines both.

Reply:
> we didn't want crashes

ossl_assert will crash in *debug* build. Just like assert would. And neither will crash in production build.

Reply:
> I thought we didn't want crashes, and that if you had an assert, there had to be a runtime test and error return, and that this led to ossl_assert which combines both.

Yes, in principle. But there are occasions where assert is fine:

````
switch {
case valid1:
    /* do something */
    break;
case valid2:
    /* do something */
    break;
default:
    assert("Shouldn't get here" == NULL);
    return 0;
}
````

There's also the case where you're in a public API function declared as void - which we can't change for compatibility reasons - so you can't return an error result.

Reply:
It's that last default case that ossl_assert handles.  Do "git grep ossl_assert" and you can see how it's used and how it combines assert with run-time check like I tried to describe.

Reply:
> how it [openssl_assert] combines assert with run-time check.

Nobody disputes that. The dispute is about whether or not styling guide should actually say "Prefer ossl_assert over assert." Of course one can argue that this formulation doesn't preclude use of assert (when appropriate), but I'd say we would be better off if wording in a *guide* is more accurate. And us having this very conversation only confirms this suggestion.

Reply:
Does "Consider using ossl_assert instead of assert" work for people?

Reply:
WFM

Reply:
As initially said, I don't see relation between ossl_assert and assert as "instead". In sense that it's not *always* the case that ossl_assert would be *more* appropriate. If anything, ossl_assert is rather replacement for normal `if (condition) handler`, except that it will crash in debug build taking the snapshot of hard-to-reproduce case. But here is question in the context. In the essence we are commenting on to-do list for updates. But what about the actual outcome? Does this to-do item mean that there will be "use of assertions" paragraph? If so, it might be meaningful to discuss *that*.

Reply:
I am trying to put down what we think is the right thing to say.  If we don't have agreement, we can just remove this item.

Reply:
There seems at least agreement that we shouldn't use OPENSSL_assert(). There is also agreement that we can use ossl_assert() and assert(). We should say at least that. I do think that we should, in most cases, prefer ossl_assert() over assert() - but it sounds like that will be difficult to agree - so we could just drop that bit of it.

Reply:
??? I'm not suggesting to remove the item. just don't see place for "ossl_assert *instead* of assert". Absolutely not if it appears as single sentence in the middle of guide. I rather expect that there will be a paragraph, i.e. multiple sentences. One can even wonder if it's worth clarifying the historical background, i.e. that there was OPENSSL_assert and that we are omitting it in favour of "soft" assertions, i.e. ones that crash only in debug builds.